5.5/test/test_replication.sh

```diff
@@ -0,0 +1,145 @@
+function start_openshift() {
+  docker run -d --name "openshift-origin" \
+    --privileged --net=host \
+    -v /:/rootfs:ro -v /var/run:/var/run:rw -v /sys:/sys:ro -v /var/lib/docker:/var/lib/docker:rw \
+    -v $OPENSHIFT_CONFIG_DIR:/var/lib/openshift/openshift.local.config \
+    -v /var/lib/openshift/openshift.local.volumes:/var/lib/openshift/openshift.local.volumes \
+    openshift/origin start \
+      --master-config=/var/lib/openshift/openshift.local.config/master/master-config.yaml \
+      --node-config=/var/lib/openshift/openshift.local.config/$OPENSHIFT_NODE_CONFIG \
+      --loglevel=5
+}
+
+function dump_logs() {
+  docker logs openshift-origin
+}
+
+#
+# Helper functions to run commands inside of the OpenShift Docker container.
+#
+function run() {
+  docker exec openshift-origin /bin/bash -c "$@"
+}
+
+function run_interactive() {
+  docker exec -i openshift-origin /bin/bash -c "$@"
+}
+
+
+# Main
+setup_dns
+start_openshift
+set +x
+echo "Waiting for OpenShift to start ..."
+wait_for_url_timed "https://${HOST_DOCKER_IP}:8443/healthz" "" 90*TIME_SEC >/dev/null
+set -x
+
+# Install mysql command line tool so we can make test queries.
+run "yum -y install mysql"
+
+# Create new project.
+run "oc new-project replication"
+
+# Allow all pods to run as privileged, so NFS server works.
+# FIXME: There doesn't seem to be a nice way to simply add a new user to an SCC
+run "oc patch scc privileged -p '{\"users\":[\"system:serviceaccount:openshift-infra:build-controller\",\"system:serviceaccount:replication:default\"]}'"
+
+# Create Persistent Volumes backed by NFS.
+cat examples/replica/nfs-pv-provider.json | run_interactive "oc process -f - | oc create -f -"
+
+# Now create MySQL replica scenario.
+cat examples/replica/mysql_replica.json | run_interactive "oc process -f - | oc create -f -"
+
+# Wait until master and slave are up.
+#set +x
+echo "Waiting for MySQL Master and Replica to come online"
+trap dump_logs ERR
+wait_for_url_timed "mysql-master.replication.svc.cluster.local:3306" "" 5*TIME_MIN >/dev/null
+wait_for_url_timed "mysql-slave.replication.svc.cluster.local:3306" "" 1*TIME_MIN >/dev/null
+set -x
+
+# Get environment variables from master, so we know the passwords.
+export $(run "oc env --list dc mysql-master | grep -v '^#'")
+
+#
+# Tests start here.
+#
+
+# Create table in master.
+run_interactive "mysql -h mysql-master.replication.svc.cluster.local -u${MYSQL_USER} -p${MYSQL_PASSWORD} ${MYSQL_DATABASE}" <<EOF
+CREATE TABLE tbl (col1 VARCHAR(20), col2 VARCHAR(20));
+INSERT INTO tbl VALUES ('foo1', 'bar1');
+EOF
+
+sleep 2
+
+# FIXME: Should log in as normal user, doesn't work right now.
+# Verify that values are replicated in slave.
+SLAVE_POD_NAME=$(run "oc get pods | grep mysql-slave | cut -f 1 -d ' '" | head -n 1)
+run "oc exec $SLAVE_POD_NAME -- /bin/bash -c 'mysql -uroot ${MYSQL_DATABASE} -e \"SELECT * FROM tbl;\"'" | grep -q foo1
+
+echo "All tests finished successfully"
```

Comment:
Hmm, this is going to be a pickle. I usually cleaned up with `docker stop $(docker ps -q) ; docker rm -f $(docker ps -qa)`. It is difficult to shut down containers that OS launched, because we do not know their name and they keep running after stopping OS. Only thing I see in the openshift-origin container is the following line:
```
2015-08-04T08:03:56.902211104Z 2015/08/04 08:03:56 received terminated signal, shutting down...
```
Maybe this is a bug and OS should properly shutdown everything?